pkg/flow/source.go

```diff
@@ -0,0 +1,152 @@
+package flow
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/grafana/agent/pkg/river/ast"
+	"github.com/grafana/agent/pkg/river/diag"
+	"github.com/grafana/agent/pkg/river/parser"
+)
+
+// Source holds the contents of a parsed Flow source.
+type Source struct {
+	sourceMap map[string][]byte
+	hash      [sha256.Size]byte // Hash of all files in sourceMap sorted by name.
+
+	// components holds the list of raw River AST blocks describing components.
+	// The Flow controller can interpret them.
+	components   []*ast.BlockStmt
+	configBlocks []*ast.BlockStmt
+}
+
+// ParseSource parses the River contents specified by bb into a Source. name
+// should be the name of the source used for reporting errors.
+//
+// bb must not be modified after passing to ParseSource.
+func ParseSource(name string, bb []byte) (*Source, error) {
+	node, err := parser.ParseFile(name, bb)
+	if err != nil {
+		return nil, err
+	}
+
+	// Look for predefined non-components blocks (i.e., logging), and store
+	// everything else into a list of components.
+	//
+	// TODO(rfratto): should this code be brought into a helper somewhere? Maybe
```

Comment:
Not for this PR since I noticed after the fact this is really moved code, but I agree the logic for what blocks are config blocks should be moved. More specifically, there is overlap in pkg/flow/internal/controller/config.go now as well with the constants for what blocks are config blocks.